src/pyroboplan/ik/nullspace_components.py

```diff
@@ -74,3 +77,72 @@ def joint_center_nullspace_component(model, q, gain=1.0):
     """
     joint_center_positions = 0.5 * (model.lowerPositionLimit + model.upperPositionLimit)
     return gain * (joint_center_positions - q)
+
+
+def collision_avoidance_nullspace_component(
+    model,
+    data,
+    collision_model,
+    collision_data,
+    q,
+    dist_padding=0.05,
+    gain=1.0,
+):
+    """
+    Returns a collision avoidance nullspace component.
+
+    These calculations are based off the following resources:
+       * https://typeset.io/pdf/a-collision-free-mpc-for-whole-body-dynamic-locomotion-and-1l6itpfk.pdf
+       * https://laas.hal.science/hal-04425002
+
+    Parameters
+    ----------
+        model : `pinocchio.Model`
+            The model from which to generate a random state.
+        data : `pinocchio.Data`
+            The model data to use for collision distance checks.
+        collision_model : `pinocchio.GeometryModel`
+            The model with which to check collision distances.
+        collision_data : `pinocchio.GeometryData`
+            The collision model data to use for collision distance checks.
+        q : array-like
+            The joint configuration for the model.
+        dist_padding : float
+            The distance padding, in meters, on the collision distances.
+            For example, a distance padding of 0.1 means collisions have an influence 10 cm away from actual collision..
+        gain : float, optional
+            A gain to modify the relative weight of this term.
+
+    Returns
+    -------
+        array-like
+            An array containing the collision avoidance nullspace terms.
+    """
+    coll_component = np.zeros_like(model.lowerPositionLimit)
+
+    # Find all the collision distances at the current state.
+    pinocchio.framesForwardKinematics(model, data, q)
+    pinocchio.computeCollisions(model, data, collision_model, collision_data, q, False)
+    pinocchio.computeDistances(model, data, collision_model, collision_data, q)
+
+    # For each collision pair within a distance threshold, calculate its collision Jacobians
+    # and use them to push the corresponding joint values away from collision.
+    for idx in range(len(collision_model.collisionPairs)):
+        dist = collision_data.distanceResults[idx].min_distance
+        if dist > dist_padding:
+            continue
+
+        distance_vec, Jcoll1, Jcoll2 = calculate_collision_vector_and_jacobians(
+            model, collision_model, data, collision_data, idx, q
+        )
+
+        # Normalize the distance vector and add this collision pair to the nullspace component.
+        dist_norm = np.linalg.norm(distance_vec)
+        if dist_norm > 1e-12:
```

Comment:
`np.finfo(float).eps`?

Reply:
I don't want to use `eps` necessarily, but I think at least making a constant is appropriate.